File: packages/shared/src/types/character.ts
export interface Character {
  id: string
  name: string
  moduleId: string
  ruleSystem: 'coc' | 'dnd'
  isPlayer: boolean
  baseAttributes: Record<string, number>
  skills: Record<string, number>
  systemData: CoCSystemData | Record<string, unknown>
  backstory: string
  status: 'active' | 'dead' | 'incapacitated'
  createdAt: string
  updatedAt: string
}

export interface CoCSystemData {
  sanity: { current: number; max: number }
  hitPoints: { current: number; max: number }
  magicPoints: { current: number; max: number }
  luck: number
  age: number
  occupation: string
}

File: server/app/schemas/character.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel


class CharacterCreate(BaseModel):
    name: str
    module_id: str
    rule_system: str
    is_player: bool = True
    base_attributes: dict[str, int] = {}
    skills: dict[str, int] = {}
    system_data: dict = {}
    backstory: str = ""


class CharacterRead(BaseModel):
    id: str
    name: str
    module_id: str | None
    rule_system: str
    is_player: bool
    base_attributes: dict
    skills: dict
    system_data: dict
    backstory: str
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CharacterUpdate(BaseModel):
    name: str | None = None
    base_attributes: dict | None = None
    skills: dict | None = None
    system_data: dict | None = None
    backstory: str | None = None
    status: str | None = None


class RollAttributesResponse(BaseModel):
    sets: list[dict[str, int]]
